src/inventorymanager.cpp

```diff
@@ -691,72 +691,102 @@ ICraftAction::ICraftAction(std::istream &is)
+		infostream << "ICraftAction::apply(): FAIL: craftresult list not found: "
+				<< "craft_inv=\"" << craft_inv.dump() << "\"" << std::endl;
 		return;
 	}
-	if(list_craftresult->getSize() < 1){
-		infostream<<"ICraftAction::apply(): FAIL: craftresult list too short: "
-				<<"craft_inv=\""<<craft_inv.dump()<<"\""<<std::endl;
+	if (list_craftresult->getSize() < 1) {
+		infostream << "ICraftAction::apply(): FAIL: craftresult list too short: "
+				<< "craft_inv=\"" << craft_inv.dump() << "\"" << std::endl;
 		return;
 	}
 
 	ItemStack crafted;
 	ItemStack craftresultitem;
 	int count_remaining = count;
-	getCraftingResult(inv_craft, crafted, false, gamedef);
+	std::vector<ItemStack> output_replacements;
+	getCraftingResult(inv_craft, crafted, output_replacements, false, gamedef);
 	PLAYER_TO_SA(player)->item_CraftPredict(crafted, player, list_craft, craft_inv);
 	bool found = !crafted.empty();
 
-	while(found && list_craftresult->itemFits(0, crafted))
-	{
+	while (found && list_craftresult->itemFits(0, crafted)) {
 		InventoryList saved_craft_list = *list_craft;
-		
+
+		std::vector<ItemStack> temp;
 		// Decrement input and add crafting output
-		getCraftingResult(inv_craft, crafted, true, gamedef);
+		getCraftingResult(inv_craft, crafted, temp, true, gamedef);
 		PLAYER_TO_SA(player)->item_OnCraft(crafted, player, &saved_craft_list, craft_inv);
 		list_craftresult->addItem(0, crafted);
 		mgr->setInventoryModified(craft_inv);
 
-		actionstream<<player->getDescription()
-				<<" crafts "
-				<<crafted.getItemString()
-				<<std::endl;
+		// Add the new replacements to the list
+		IItemDefManager *itemdef = gamedef->getItemDefManager();
+		for (std::vector<ItemStack>::iterator it = temp.begin();
+				it != temp.end(); it++) {
+			for (std::vector<ItemStack>::iterator jt = output_replacements.begin();
+					jt != output_replacements.end(); jt++) {
+				if (it->name == jt->name) {
+					*it = jt->addItem(*it, itemdef);
+					if (it->empty())
+						continue;
+				}
+			}
+			output_replacements.push_back(*it);
+		}
+
+		actionstream << player->getDescription()
+				<< " crafts "
+				<< crafted.getItemString()
+				<< std::endl;
 
 		// Decrement counter
-		if(count_remaining == 1)
+		if (count_remaining == 1)
 			break;
-		else if(count_remaining > 1)
+		else if (count_remaining > 1)
 			count_remaining--;
 
 		// Get next crafting result
-		found = getCraftingResult(inv_craft, crafted, false, gamedef);
+		found = getCraftingResult(inv_craft, crafted, temp, false, gamedef);
 		PLAYER_TO_SA(player)->item_CraftPredict(crafted, player, list_craft, craft_inv);
 		found = !crafted.empty();
 	}
 
+	// Put the replacements in the inventory or drop them on the floor, if
+	// the invenotry is full
+	for (std::vector<ItemStack>::iterator it = output_replacements.begin();
+			it != output_replacements.end(); it++) {
+		if (list_main)
+			*it = list_main->addItem(*it);
+		while (!it->empty()) {
+			PLAYER_TO_SA(player)->item_OnDrop(*it, player,
```

Comment:
This can hang the server if the stack isn't decremented by the on_drop function. How would this be done the correct way? Is there another way to drop stuff in the world without going to lua?